functions/sendgrid/index.js

```diff
@@ -13,101 +13,345 @@
+      }
+    }
+  }
+}
+// [END fixNames]
+
+// [START webhook]
+/**
+ * Receive a webhook from SendGrid.
+ *
+ * See https://sendgrid.com/docs/API_Reference/Webhooks/event.html
+ *
+ * @param {Object} req Cloud Function request context.
+ * @param {Object} res Cloud Function response context.
+ */
+exports.sendgridWebhook = function sendgridWebhook (req, res) {
+  try {
+    if (req.method !== 'POST') {
+      var error = new Error('Only POST requests are accepted');
+      error.code = 405;
+      throw error;
+    }
+
+    verifyWebhook(req.get('authorization') || '');
+
+    var events = req.body || [];
+
+    fixNames(events);
+
+    // Generate newline-delimite JSON
+    // See https://cloud.google.com/bigquery/data-formats#json_format
+    var json = events.map(function (event) {
+      return JSON.stringify(event);
+    }).join('\n');
+
+    // Upload a new file to Cloud Storage if we have events to save
+    if (json.length) {
+      var bucketName = config.EVENT_BUCKET;
+      var filename = uuid.v4() + '.json';
+      var file = storage.bucket(bucketName).file(filename);
+
+      console.log('Saving events to ' + filename + ' in bucket ' + bucketName);
+
+      return file.save(json, function (err) {
+        if (err) {
+          console.error(err);
+          return res.status(500).end();
+        }
+        console.log('JSON written to ' + filename);
+        return res.status(200).end();
+      });
+    }
+
+    return res.status(200).end();
+  } catch (err) {
+    console.error(err);
+    return res.status(err.code || 500).send(err.message);
+  }
+};
+// [END webhook]
+
+// [START getTable]
+/**
+ * Helper method to get a handle on a BigQuery table. Automatically creates the
+ * dataset and table if necessary.
+ *
+ * @param {Function} callback Callback function.
+ */
+function getTable (callback) {
+  var dataset = bigquery.dataset(config.DATASET);
+  return dataset.get({
+    autoCreate: true
+  }, function (err, dataset) {
+    if (err) {
+      return callback(err);
+    }
+    var table = dataset.table(config.TABLE);
+    return table.get({
+      autoCreate: true
+    }, function (err, table) {
+      if (err) {
+        return callback(err);
+      }
+      return callback(null, table);
+    });
+  });
+}
+// [END getTable]
+
+// [START load]
+/**
+ * Cloud Function triggered by Cloud Storage when a file is uploaded.
+ *
+ * @param {Object} context Cloud Function context.
+ * @param {Function} context.success Success callback.
+ * @param {Function} context.failure Failure callback.
+ * @param {Object} data Request data, in this case an object provided by Cloud Storage.
+ * @param {string} data.bucket Name of the Cloud Storage bucket.
+ * @param {string} data.name Name of the file.
+ * @param {string} [data.timeDeleted] Time the file was deleted if this is a deletion event.
+ * @see https://cloud.google.com/storage/docs/json_api/v1/objects#resource
+ */
+exports.sendgridLoad = function sendgridLoad (context, data) {
```

Comment:
So I think one thing that somewhat nags me about these larger functions - I would prefer two functions:

```
exports.onCloudStorage(context, data) {
    loadDataToBigQuery(...);
}
```

but that's just like my opinion, man. It's up to you.